// @mainnet-cash/contract/dist/module/WrappedProvider.js
import { ElectrumNetworkProvider } from "cashscript";
import { Network, } from "mainnet-js";
export const toCashScript = (utxo) => ({
    satoshis: BigInt(utxo.satoshis),
    txid: utxo.txid,
    vout: utxo.vout,
    token: utxo.token
        ? {
            amount: utxo.token?.amount ? BigInt(utxo.token.amount) : 0n,
            category: utxo.token?.tokenId,
            nft: utxo.token?.capability || utxo.token?.commitment
                ? {
                    capability: utxo.token?.capability,
                    commitment: utxo.token?.commitment,
                }
                : undefined,
        }
        : undefined,
});
export const toMainnet = (utxo) => ({
    satoshis: Number(utxo.satoshis),
    txid: utxo.txid,
    vout: utxo.vout,
    token: utxo.token
        ? {
            amount: utxo.token?.amount ? BigInt(utxo.token?.amount) : undefined,
            tokenId: utxo.token?.category,
            capability: utxo.token?.nft?.capability,
            commitment: utxo.token?.nft?.commitment,
        }
        : undefined,
});
export default class WrappedProvider extends ElectrumNetworkProvider {
    constructor(mainnetProvider) {
        super(mainnetProvider.network === Network.TESTNET
            ? "chipnet"
            : mainnetProvider.network, {
            electrum: mainnetProvider.electrum,
            manualConnectionManagement: false,
        });
        this.mainnetProvider = mainnetProvider;
        this.performRequest = async (name, ...parameters) => {
            return this.mainnetProvider.performRequest(name, ...parameters);
        };
    }
    /**
     * Retrieve all UTXOs (confirmed and unconfirmed) for a given address.
     * @param address The CashAddress for which we wish to retrieve UTXOs.
     * @returns List of UTXOs spendable by the provided address.
     */
    async getUtxos(address) {
        return (await this.mainnetProvider.getUtxos(address)).map(toCashScript);
    }
}
