// cashscript/dist/interfaces.js
export function isUnlockableUtxo(utxo) {
    return 'unlocker' in utxo;
}
export function isStandardUnlockableUtxo(utxo) {
    return isStandardUnlocker(utxo.unlocker);
}
export function isContractUnlocker(unlocker) {
    return 'contract' in unlocker;
}
export function isP2PKHUnlocker(unlocker) {
    return 'template' in unlocker;
}
export function isStandardUnlocker(unlocker) {
    return isContractUnlocker(unlocker) || isP2PKHUnlocker(unlocker);
}
export function isPlaceholderUnlocker(unlocker) {
    return 'placeholder' in unlocker;
}
export var SignatureAlgorithm;
(function (SignatureAlgorithm) {
    SignatureAlgorithm[SignatureAlgorithm["ECDSA"] = 0] = "ECDSA";
    SignatureAlgorithm[SignatureAlgorithm["SCHNORR"] = 1] = "SCHNORR";
})(SignatureAlgorithm || (SignatureAlgorithm = {}));
export var HashType;
(function (HashType) {
    HashType[HashType["SIGHASH_ALL"] = 1] = "SIGHASH_ALL";
    HashType[HashType["SIGHASH_NONE"] = 2] = "SIGHASH_NONE";
    HashType[HashType["SIGHASH_SINGLE"] = 3] = "SIGHASH_SINGLE";
    HashType[HashType["SIGHASH_UTXOS"] = 32] = "SIGHASH_UTXOS";
    HashType[HashType["SIGHASH_ANYONECANPAY"] = 128] = "SIGHASH_ANYONECANPAY";
})(HashType || (HashType = {}));
// Weird setup to allow both Enum parameters, as well as literal strings
// https://stackoverflow.com/questions/51433319/typescript-constructor-accept-string-for-enum
const literal = (l) => l;
export const Network = {
    MAINNET: literal('mainnet'),
    TESTNET3: literal('testnet3'),
    TESTNET4: literal('testnet4'),
    CHIPNET: literal('chipnet'),
    MOCKNET: literal('mocknet'),
    REGTEST: literal('regtest'),
};
export const VmTarget = {
    BCH_2023_05: literal('BCH_2023_05'),
    BCH_2025_05: literal('BCH_2025_05'),
    BCH_2026_05: literal('BCH_2026_05'),
    BCH_SPEC: literal('BCH_SPEC'),
};
