// cashscript/dist/walletconnect-utils.js
import { isContractUnlocker } from './interfaces.js';
import { scriptToBytecode } from '@cashscript/utils';
import { cashAddressToLockingBytecode } from '@bitauth/libauth';
export function getWcContractInfo(input) {
    // If the input does not have a contract unlocker, return an empty object
    if (!(isContractUnlocker(input.unlocker)))
        return {};
    const contract = input.unlocker.contract;
    const abiFunctionName = input.unlocker.abiFunction?.name;
    const abiFunction = contract.artifact.abi.find(abi => abi.name === abiFunctionName);
    if (!abiFunction) {
        throw new Error(`ABI function ${abiFunctionName} not found in contract artifact`);
    }
    const wcContractObj = {
        contract: {
            abiFunction: abiFunction,
            redeemScript: scriptToBytecode(contract.redeemScript),
            artifact: contract.artifact,
        },
    };
    return wcContractObj;
}
export const placeholderSignature = () => Uint8Array.from(Array(65));
export const placeholderPublicKey = () => Uint8Array.from(Array(33));
export const placeholderP2PKHUnlocker = (userAddress) => {
    const decodeAddressResult = cashAddressToLockingBytecode(userAddress);
    if (typeof decodeAddressResult === 'string') {
        throw new Error(`Invalid address: ${decodeAddressResult}`);
    }
    const lockingBytecode = decodeAddressResult.bytecode;
    return {
        generateLockingBytecode: () => lockingBytecode,
        generateUnlockingBytecode: () => Uint8Array.from(Array(0)),
        placeholder: true,
    };
};
